File: tools/plots/x13plot.py
"""X13cpp plot-family prototype (matplotlib, light mode).

Consumes oracle/golden bundles via the x13compare parsers. This is the design
prototype for the eventual x13cpp Python package plotting API; the same visual
system will be mirrored in ggplot2 for the R package.

Design system (see project plan / dataviz method):
  surface #fcfcfb, ink #0b0b0b / #52514e / muted #898781,
  grid #e1e0d9, baseline #c3c2b7,
  series slots: 1 blue #2a78d6, 2 green #008300, 3 magenta #e87ba4,
                6 orange #eb6834, 8 red #e34948
  status: good #0ca30c, warning #fab219, critical #d03b3b
"""
from __future__ import annotations

import os
import sys
from datetime import datetime

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import numpy as np

_HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.join(_HERE, "..", "..", "tests", "compare"))
from x13compare import parse_save, parse_udg  # noqa: E402

# --------------------------------------------------------------------------- #
# Palette / chrome
# --------------------------------------------------------------------------- #
SURFACE = "#fcfcfb"
INK = "#0b0b0b"
INK2 = "#52514e"
MUTED = "#898781"
GRID = "#e1e0d9"
BASELINE = "#c3c2b7"

BLUE = "#2a78d6"       # slot 1 — primary series (SA)
BLUE_DARK = "#1c5cab"  # sequential 550 — emphasis of the blue family
BLUE_LIGHT = "#cde2fb"  # sequential 100 — fills
BLUE_MID = "#9ec5f4"    # sequential 200
GREEN = "#008300"      # slot 2 — trend
RED = "#e34948"        # slot 8 — flagged points
CRITICAL = "#d03b3b"   # status — failing stat
GOOD_TEXT = "#006300"  # success text on light

FONT = ["Segoe UI", "DejaVu Sans", "sans-serif"]

RC = {
    "figure.facecolor": SURFACE,
    "axes.facecolor": SURFACE,
    "savefig.facecolor": SURFACE,
    "font.family": "sans-serif",
    "font.sans-serif": FONT,
    "text.color": INK,
    "axes.edgecolor": BASELINE,
    "axes.labelcolor": INK2,
    "axes.titlecolor": INK,
    "xtick.color": MUTED,
    "ytick.color": MUTED,
    "xtick.labelsize": 9,
    "ytick.labelsize": 9,
    "axes.labelsize": 10,
    "axes.grid": True,
    "grid.color": GRID,
    "grid.linewidth": 0.8,
    "axes.axisbelow": True,
    "axes.spines.top": False,
    "axes.spines.right": False,
    "legend.frameon": False,
    "legend.fontsize": 9,
    "axes.titlesize": 13,
    "axes.titleweight": "bold",
    "axes.titlelocation": "left",
}


def _apply_style():
    plt.rcParams.update(RC)


def _finish(ax):
    ax.grid(axis="x", visible=False)
    for s in ("bottom", "left"):
        ax.spines[s].set_linewidth(0.9)


def _title(fig, title, subtitle):
    fig.suptitle(title, x=0.06, y=0.975, ha="left", fontsize=14,
                 fontweight="bold", color=INK)
    fig.text(0.06, 0.905, subtitle, ha="left", fontsize=10, color=INK2)


# --------------------------------------------------------------------------- #
# Bundle access
# --------------------------------------------------------------------------- #
def _period_to_date(period: str) -> datetime:
    p = str(period)
    if len(p) == 6:                       # YYYYMM
        return datetime(int(p[:4]), int(p[4:6]), 1)
    if len(p) == 4:                       # YYYY
        return datetime(int(p), 1, 1)
    raise ValueError(f"unrecognized period {period!r}")


class Bundle:
    """Read save tables / udg from an oracle output bundle directory."""

    def __init__(self, path: str):
        self.path = path
        self.base = None
        for f in os.listdir(path):
            if f.endswith(".out"):
                self.base = f[:-4]
        if self.base is None:
            raise FileNotFoundError(f"no .out in {path}")

    def table(self, code: str):
        """Return (dates, values) for a saved table as (list, 2-D array).

        Prefers the multi-column ``rows`` mapping (e.g. .fct has
        forecast/lowerci/upperci). X-13 encodes missing values as the
        sentinel -999.0 — mapped to NaN here.
        """
        f = os.path.join(self.path, f"{self.base}.{code}")
        parsed = parse_save.parse_save(open(f, encoding="utf-8").read())
        rows = getattr(parsed, "rows", None)
        if rows:
            items = sorted(rows.items())
        else:
            items = sorted(parsed.values.items())
        dates = [_period_to_date(k) for k, _ in items]
        vals = np.array([v if isinstance(v, (list, tuple)) else [v]
                         for _, v in items], dtype=float)
        vals[np.isclose(vals, -999.0)] = np.nan
        return dates, vals

    def udg(self) -> dict:
        f = os.path.join(self.path, f"{self.base}.udg")
        return parse_udg.parse_udg(open(f, encoding="utf-8").read())

    @property
    def period(self) -> int:
        try:
            return int(self.udg().get("freq", 12))
        except Exception:
            return 12


MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
          "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


# --------------------------------------------------------------------------- #
# 1. Overview: original, SA, trend
# --------------------------------------------------------------------------- #
def plot_overview(bundle: Bundle, title, subtitle, out_png):
    _apply_style()
    d_o, v_o = bundle.table("a1")
    d_sa, v_sa = bundle.table("d11")
    d_tr, v_tr = bundle.table("d12")

    fig, ax = plt.subplots(figsize=(9.6, 4.6), dpi=150)
    fig.subplots_adjust(top=0.86, left=0.09, right=0.97, bottom=0.11)
    ax.plot(d_o, v_o[:, 0], color=MUTED, lw=1.1, label="Original (A1)")
    ax.plot(d_sa, v_sa[:, 0], color=BLUE, lw=1.9,
            label="Seasonally adjusted (D11)")
    ax.plot(d_tr, v_tr[:, 0], color=GREEN, lw=1.6, label="Trend-cycle (D12)")
    ax.legend(loc="upper left", ncols=3, bbox_to_anchor=(0.0, 1.02))
    ax.yaxis.set_major_formatter(
        matplotlib.ticker.FuncFormatter(lambda v, _: f"{v:,.0f}"))
    ax.xaxis.set_major_locator(mdates.AutoDateLocator())
    _finish(ax)
    _title(fig, title, subtitle)
    fig.savefig(out_png)
    plt.close(fig)


# --------------------------------------------------------------------------- #
# 2. Seasonal factor subseries (per-period panels)
# --------------------------------------------------------------------------- #
def plot_seasonal_subseries(bundle: Bundle, title, subtitle, out_png):
    _apply_style()
    d10_d, d10_v = bundle.table("d10")
    per = bundle.period
    months = [d.month if per == 12 else (d.month - 1) // 3 + 1 for d in d10_d]
    years = [d.year for d in d10_d]
    labels = MONTHS if per == 12 else [f"Q{i}" for i in range(1, 5)]

    fig, axes = plt.subplots(1, per, figsize=(11.5, 4.2), dpi=150,
                             sharey=True)
    fig.subplots_adjust(top=0.80, left=0.07, right=0.985, bottom=0.10,
                        wspace=0.06)
    mean_all = np.nanmean(d10_v[:, 0])
    if mean_all > 50:          # multiplicative, percent scale (~100)
        ref = 100.0
    elif 0.5 < mean_all < 2:   # multiplicative, ratio scale (~1)
        ref = 1.0
    else:                      # additive (~0)
        ref = 0.0
    lo = np.nanmin(d10_v[:, 0]); hi = np.nanmax(d10_v[:, 0])
    pad = 0.08 * (hi - lo) if hi > lo else 0.01
    axes[0].set_ylim(min(lo, ref) - pad, max(hi, ref) + pad)
    for m in range(1, per + 1):
        ax = axes[m - 1]
        yy = [v for mm, v in zip(months, d10_v[:, 0]) if mm == m]
        xx = [y for mm, y in zip(months, years) if mm == m]
        ax.axhline(ref, color=BASELINE, lw=0.9)
        ax.plot(xx, yy, color=BLUE, lw=1.6)
        ax.axhline(np.nanmean(yy), color=BLUE_DARK, lw=1.0, ls=(0, (4, 3)))
        ax.set_title(labels[m - 1], fontsize=9, fontweight="normal",
                     color=INK2, loc="center")
        ax.set_xticks([])
        ax.grid(visible=False)
        for s in ("bottom", "left"):
            ax.spines[s].set_linewidth(0.9)
        if m > 1:
            ax.spines["left"].set_visible(False)
            ax.tick_params(left=False)
    _title(fig, title, subtitle)
    fig.savefig(out_png)
    plt.close(fig)


# --------------------------------------------------------------------------- #
# 3. SI ratios with replacement + seasonal factors
# --------------------------------------------------------------------------- #
def plot_si_ratios(bundle: Bundle, title, subtitle, out_png):
    _apply_style()
    d8_d, d8_v = bundle.table("d8")     # unmodified SI
    d9_d, d9_v = bundle.table("d9")     # replaced SI (sparse: mostly missing)
    d10_d, d10_v = bundle.table("d10")  # seasonal factors
    per = bundle.period
    labels = MONTHS if per == 12 else [f"Q{i}" for i in range(1, 5)]

    d9_map = {d: v for d, v in zip(d9_d, d9_v[:, 0]) if np.isfinite(v)}

    fig, axes = plt.subplots(1, per, figsize=(11.5, 4.2), dpi=150,
                             sharey=True)
    fig.subplots_adjust(top=0.80, left=0.07, right=0.985, bottom=0.10,
                        wspace=0.06)
    fin = d8_v[:, 0][np.isfinite(d8_v[:, 0])]
    lo, hi = fin.min(), fin.max()
    pad = 0.08 * (hi - lo)
    axes[0].set_ylim(lo - pad, hi + pad)
    for m in range(1, per + 1):
        ax = axes[m - 1]
        si = [(d.year, v) for d, v in zip(d8_d, d8_v[:, 0])
              if (d.month if per == 12 else (d.month - 1) // 3 + 1) == m]
        sf = [(d.year, v) for d, v in zip(d10_d, d10_v[:, 0])
              if (d.month if per == 12 else (d.month - 1) // 3 + 1) == m]
        rep = [(d.year, v) for d, v in d9_map.items()
               if (d.month if per == 12 else (d.month - 1) // 3 + 1) == m]
        ax.scatter([x for x, _ in si], [y for _, y in si], s=7, color=MUTED,
                   linewidths=0)
        ax.plot([x for x, _ in sf], [y for _, y in sf], color=BLUE, lw=1.6)
        if rep:
            ax.scatter([x for x, _ in rep], [y for _, y in rep], s=16,
                       color=RED, linewidths=0, zorder=3)
        ax.set_title(labels[m - 1], fontsize=9, fontweight="normal",
                     color=INK2, loc="center")
        ax.set_xticks([])
        ax.grid(visible=False)
        if m > 1:
            ax.spines["left"].set_visible(False)
            ax.tick_params(left=False)
    _title(fig, title, subtitle)
    fig.savefig(out_png)
    plt.close(fig)


# --------------------------------------------------------------------------- #
# 4. Forecast fan
# --------------------------------------------------------------------------- #
def plot_forecast(bundle: Bundle, title, subtitle, out_png, history=48):
    _apply_style()
    a1_d, a1_v = bundle.table("a1")
    f_d, f_v = bundle.table("fct")      # cols: forecast, lower, upper

    fig, ax = plt.subplots(figsize=(9.6, 4.6), dpi=150)
    fig.subplots_adjust(top=0.86, left=0.09, right=0.97, bottom=0.11)
    hd, hv = a1_d[-history:], a1_v[-history:, 0]
    ax.plot(hd, hv, color=BLUE, lw=1.9, label="Observed")
    if f_v.shape[1] >= 3:
        ax.fill_between(f_d, f_v[:, 1], f_v[:, 2], color=BLUE_LIGHT,
                        lw=0, label="95% interval")
    ax.plot(f_d, f_v[:, 0], color=BLUE_DARK, lw=1.9, ls=(0, (5, 3)),
            label="Forecast")
    ax.axvline(hd[-1], color=BASELINE, lw=0.9)
    ax.legend(loc="upper left", ncols=3, bbox_to_anchor=(0.0, 1.02))
    _finish(ax)
    _title(fig, title, subtitle)
    fig.savefig(out_png)
    plt.close(fig)


# --------------------------------------------------------------------------- #
# 5. M-statistics dashboard
# --------------------------------------------------------------------------- #
M_DESC = {
    "m01": "Irregular contribution",
    "m02": "Irregular in trend",
    "m03": "I/C ratio",
    "m04": "Irregular autocorrelation",
    "m05": "MCD months",
    "m06": "I/S yearly",
    "m07": "Moving seasonality",
    "m08": "Seasonal fluctuation",
    "m09": "Seasonal linear movement",
    "m10": "Recent seasonal fluct.",
    "m11": "Recent seasonal movement",
}


def plot_mstats(bundle: Bundle, title, subtitle, out_png):
    _apply_style()
    u = bundle.udg()
    keys = [f"m{i:02d}" for i in range(1, 12)]
    vals = []
    for k in keys:
        v = u.get(f"f3.{k}")
        vals.append(float(v) if v is not None else np.nan)
    q = float(u.get("f3.q", np.nan))

    fig, ax = plt.subplots(figsize=(9.6, 5.0), dpi=150)
    fig.subplots_adjust(top=0.85, left=0.30, right=0.93, bottom=0.09)
    y = np.arange(len(keys))[::-1]
    colors = [CRITICAL if v >= 1.0 else BLUE for v in vals]
    ax.barh(y, vals, height=0.62, color=colors, zorder=3)
    ax.axvline(1.0, color=INK2, lw=1.1, zorder=4)
    ax.text(1.0, len(keys) - 0.1, " acceptance limit", color=INK2,
            fontsize=8.5, va="bottom", ha="left")
    ax.set_yticks(y)
    ax.set_yticklabels([f"{k.upper()}  ·  {M_DESC[k]}" for k in keys],
                       fontsize=9, color=INK)
    for yi, v in zip(y, vals):
        if np.isfinite(v):
            ax.text(v + 0.04, yi, f"{v:.2f}", va="center", fontsize=8.5,
                    color=(CRITICAL if v >= 1.0 else INK2), zorder=5)
    ax.set_xlim(0, max(3.2, np.nanmax(vals) * 1.12))
    ax.grid(axis="y", visible=False)
    ax.grid(axis="x", visible=True)
    ax.spines["left"].set_visible(False)
    ax.tick_params(left=False)
    q_col = GOOD_TEXT if q < 1.0 else CRITICAL
    verdict = "accepted" if q < 1.0 else "not accepted"
    fig.text(0.93, 0.90, f"Q = {q:.2f}  ({verdict})", ha="right", fontsize=11,
             color=q_col, fontweight="bold")
    _title(fig, title, subtitle)
    fig.savefig(out_png)
    plt.close(fig)

File: tools/plots/R/x13plot.R
# X13cpp plot-family — R / ggplot2 mirror of tools/plots/x13plot.py
#
# Reproduces the five prototype plots (overview, seasonal subseries, SI ratios,
# forecast fan, M-stats dashboard) from oracle/golden output bundles, using the
# same light-mode design system as the matplotlib prototype.
#
# Design system:
#   surface #fcfcfb, ink #0b0b0b / #52514e, muted #898781,
#   grid #e1e0d9, baseline #c3c2b7,
#   blue #2a78d6, blue-dark #1c5cab, blue-light #cde2fb,
#   green #008300, red #e34948, critical #d03b3b, good-text #006300.

suppressPackageStartupMessages({
  library(ggplot2)
})

# --------------------------------------------------------------------------- #
# Palette / chrome
# --------------------------------------------------------------------------- #
SURFACE    <- "#fcfcfb"
INK        <- "#0b0b0b"
INK2       <- "#52514e"
MUTED      <- "#898781"
GRID       <- "#e1e0d9"
BASELINE   <- "#c3c2b7"

BLUE       <- "#2a78d6"   # primary series (SA)
BLUE_DARK  <- "#1c5cab"   # emphasis / forecast
BLUE_LIGHT <- "#cde2fb"   # interval fills
GREEN      <- "#008300"   # trend
RED        <- "#e34948"   # flagged points
CRITICAL   <- "#d03b3b"   # failing stat
GOOD_TEXT  <- "#006300"   # success text

# Prefer Segoe UI where the graphics device can resolve it; harmless elsewhere.
BASE_FONT <- "Segoe UI"
try(suppressWarnings(windowsFonts(`Segoe UI` = windowsFont("Segoe UI"))),
    silent = TRUE)

MONTHS <- c("Jan", "Feb", "Mar", "Apr", "May", "Jun",
            "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")

MISSING <- -999.0

# --------------------------------------------------------------------------- #
# Bundle access
# --------------------------------------------------------------------------- #

# Locate the base name of a bundle directory (the stem of its .out file).
bundle_base <- function(path) {
  out <- list.files(path, pattern = "\\.out$")
  if (length(out) == 0L) stop(sprintf("no .out in %s", path))
  sub("\\.out$", "", out[[1L]])
}

# Read an X-13 save table into a data frame.
#
# Save files have two header lines (date<TAB>label(s), then dashes), followed by
# YYYYMM<TAB>value(<TAB>value...) rows. The sentinel -999 marks missing values
# and is mapped to NA. Multi-column tables (e.g. .fct: forecast/lowerci/upperci)
# keep every value column, named from the header line.
read_save <- function(path, code, base = bundle_base(path)) {
  f <- file.path(path, paste0(base, ".", code))
  lines <- readLines(f, warn = FALSE)
  lines <- lines[nzchar(trimws(lines))]
  if (length(lines) < 3L) stop(sprintf("save file too short: %s", f))

  hdr <- trimws(strsplit(lines[[1L]], "\t")[[1L]])
  value_names <- hdr[-1L]

  body <- lines[-(1:2)]
  toks <- strsplit(trimws(body), "\\s+")
  ncol <- length(toks[[1L]])

  period <- vapply(toks, `[`, character(1L), 1L)
  vals <- matrix(NA_real_, nrow = length(toks), ncol = ncol - 1L)
  for (j in seq_len(ncol - 1L)) {
    vals[, j] <- as.numeric(vapply(toks, `[`, character(1L), j + 1L))
  }
  vals[abs(vals - MISSING) < 1e-6] <- NA_real_

  df <- data.frame(period = period, date = period_to_date(period),
                   stringsAsFactors = FALSE)
  vnames <- if (length(value_names) == ncol - 1L) {
    make.names(value_names, unique = TRUE)
  } else {
    if (ncol - 1L == 1L) "value" else paste0("value", seq_len(ncol - 1L))
  }
  # Canonical first-column name for single-series tables.
  if (ncol - 1L == 1L) vnames <- "value"
  for (j in seq_len(ncol - 1L)) df[[vnames[[j]]]] <- vals[, j]
  attr(df, "value_names") <- vnames
  df
}

# Convert a YYYYMM (or YYYY) period string to a Date on the first of the period.
period_to_date <- function(period) {
  period <- as.character(period)
  n <- nchar(period)
  y <- ifelse(n >= 4L, as.integer(substr(period, 1L, 4L)), NA_integer_)
  m <- ifelse(n == 6L, as.integer(substr(period, 5L, 6L)), 1L)
  as.Date(sprintf("%04d-%02d-01", y, m))
}

# Parse a .udg file ("key: value" lines) into a named list of trimmed strings.
read_udg <- function(path, base = bundle_base(path)) {
  f <- file.path(path, paste0(base, ".udg"))
  lines <- readLines(f, warn = FALSE)
  lines <- lines[grepl(":", lines, fixed = TRUE)]
  out <- list()
  for (ln in lines) {
    idx <- regexpr(":", ln, fixed = TRUE)
    key <- trimws(substr(ln, 1L, idx - 1L))
    val <- trimws(substr(ln, idx + 1L, nchar(ln)))
    if (nzchar(key)) out[[key]] <- val
  }
  out
}

udg_num <- function(udg, key, default = NA_real_) {
  v <- udg[[key]]
  if (is.null(v)) return(default)
  suppressWarnings(as.numeric(v))
}

bundle_period <- function(udg) {
  p <- udg_num(udg, "freq", 12)
  if (is.na(p)) 12L else as.integer(p)
}

# --------------------------------------------------------------------------- #
# Theme
# --------------------------------------------------------------------------- #

# Base ggplot theme implementing the design system. `time_series = TRUE` drops
# vertical grid lines and adds bottom/left axis lines (for line/scatter panels);
# `time_series = FALSE` keeps vertical grid (for the horizontal bar dashboard).
theme_x13 <- function(base_size = 11, time_series = TRUE) {
  th <- theme_minimal(base_size = base_size, base_family = BASE_FONT) +
    theme(
      plot.background   = element_rect(fill = SURFACE, colour = NA),
      panel.background  = element_rect(fill = SURFACE, colour = NA),
      legend.background = element_rect(fill = SURFACE, colour = NA),
      legend.key        = element_rect(fill = SURFACE, colour = NA),
      text              = element_text(colour = INK),
      plot.title        = element_text(colour = INK, face = "bold",
                                       size = base_size + 3, hjust = 0,
                                       margin = margin(b = 2)),
      plot.subtitle     = element_text(colour = INK2, size = base_size - 1,
                                       hjust = 0, margin = margin(b = 10)),
      plot.title.position = "plot",
      axis.title        = element_text(colour = INK2, size = base_size - 1),
      axis.text         = element_text(colour = MUTED, size = base_size - 2),
      axis.ticks        = element_line(colour = BASELINE, linewidth = 0.4),
      panel.grid.minor  = element_blank(),
      legend.title      = element_blank(),
      legend.position   = "top",
      legend.justification = "left",
      legend.direction  = "horizontal",
      legend.margin     = margin(0, 0, 0, 0),
      legend.box.spacing = unit(2, "pt"),
      strip.text        = element_text(colour = INK2, size = base_size - 2),
      plot.margin       = margin(12, 16, 10, 12)
    )
  if (time_series) {
    th <- th + theme(
      panel.grid.major.x = element_blank(),
      panel.grid.major.y = element_line(colour = GRID, linewidth = 0.4),
      axis.line.x        = element_line(colour = BASELINE, linewidth = 0.5),
      axis.line.y        = element_line(colour = BASELINE, linewidth = 0.5)
    )
  } else {
    th <- th + theme(
      panel.grid.major.y = element_blank(),
      panel.grid.major.x = element_line(colour = GRID, linewidth = 0.4)
    )
  }
  th
}

# --------------------------------------------------------------------------- #
# 1. Overview: original, SA, trend
# --------------------------------------------------------------------------- #
plot_overview <- function(bundle, title, subtitle) {
  base <- bundle_base(bundle)
  o  <- read_save(bundle, "a1",  base)
  sa <- read_save(bundle, "d11", base)
  tr <- read_save(bundle, "d12", base)

  lvl <- c("Original (A1)", "Seasonally adjusted (D11)", "Trend-cycle (D12)")
  df <- rbind(
    data.frame(date = o$date,  value = o$value,  series = lvl[1]),
    data.frame(date = sa$date, value = sa$value, series = lvl[2]),
    data.frame(date = tr$date, value = tr$value, series = lvl[3])
  )
  df$series <- factor(df$series, levels = lvl)

  ggplot(df, aes(date, value, colour = series, linewidth = series)) +
    geom_line(na.rm = TRUE) +
    scale_colour_manual(values = c(MUTED, BLUE, GREEN)) +
    scale_linewidth_manual(values = c(1.1, 1.9, 1.6) * 0.4) +
    scale_x_date(expand = expansion(mult = c(0.01, 0.01))) +
    scale_y_continuous(labels = function(v) format(v, big.mark = ",",
                                                    scientific = FALSE,
                                                    trim = TRUE)) +
    labs(title = title, subtitle = subtitle, x = NULL, y = NULL) +
    guides(linewidth = "none",
           colour = guide_legend(override.aes = list(linewidth = 1.2))) +
    theme_x13()
}

# --------------------------------------------------------------------------- #
# helpers for period indexing
# --------------------------------------------------------------------------- #
period_index <- function(dates, per) {
  m <- as.integer(format(dates, "%m"))
  if (per == 12L) m else (m - 1L) %/% 3L + 1L
}
period_labels <- function(per) {
  if (per == 12L) MONTHS else paste0("Q", 1:4)
}

# --------------------------------------------------------------------------- #
# 2. Seasonal factor subseries (per-period panels)
# --------------------------------------------------------------------------- #
plot_seasonal_subseries <- function(bundle, title, subtitle) {
  base <- bundle_base(bundle)
  udg  <- read_udg(bundle, base)
  per  <- bundle_period(udg)
  d10  <- read_save(bundle, "d10", base)

  idx <- period_index(d10$date, per)
  labs_v <- period_labels(per)
  df <- data.frame(year = as.integer(format(d10$date, "%Y")),
                   value = d10$value, idx = idx)
  df$panel <- factor(labs_v[df$idx], levels = labs_v)

  mean_all <- mean(df$value, na.rm = TRUE)
  ref <- if (mean_all > 50) 100.0 else if (mean_all > 0.5 && mean_all < 2) 1.0 else 0.0
  lo <- min(df$value, na.rm = TRUE); hi <- max(df$value, na.rm = TRUE)
  pad <- if (hi > lo) 0.08 * (hi - lo) else 0.01
  ylim <- c(min(lo, ref) - pad, max(hi, ref) + pad)

  means <- aggregate(value ~ panel, df, mean, na.rm = TRUE)

  ggplot(df, aes(year, value)) +
    geom_hline(yintercept = ref, colour = BASELINE, linewidth = 0.4) +
    geom_hline(data = means, aes(yintercept = value),
               colour = BLUE_DARK, linewidth = 0.4, linetype = "22") +
    geom_line(colour = BLUE, linewidth = 0.6, na.rm = TRUE) +
    facet_wrap(~panel, nrow = 1, scales = "free_x") +
    coord_cartesian(ylim = ylim) +
    labs(title = title, subtitle = subtitle, x = NULL, y = NULL) +
    theme_x13() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),
          axis.line.x = element_blank(),
          panel.spacing.x = unit(4, "pt"))
}

# --------------------------------------------------------------------------- #
# 3. SI ratios with replacement + seasonal factors
# --------------------------------------------------------------------------- #
plot_si_ratios <- function(bundle, title, subtitle) {
  base <- bundle_base(bundle)
  udg  <- read_udg(bundle, base)
  per  <- bundle_period(udg)
  d8   <- read_save(bundle, "d8",  base)   # unmodified SI
  d9   <- read_save(bundle, "d9",  base)   # replaced SI (sparse)
  d10  <- read_save(bundle, "d10", base)   # seasonal factors
  labs_v <- period_labels(per)

  mk <- function(sv, dates) {
    data.frame(year = as.integer(format(dates, "%Y")), value = sv,
               idx = period_index(dates, per))
  }
  si  <- mk(d8$value,  d8$date)
  sf  <- mk(d10$value, d10$date)
  rep <- mk(d9$value,  d9$date)
  rep <- rep[is.finite(rep$value), , drop = FALSE]

  si$panel  <- factor(labs_v[si$idx],  levels = labs_v)
  sf$panel  <- factor(labs_v[sf$idx],  levels = labs_v)
  if (nrow(rep)) rep$panel <- factor(labs_v[rep$idx], levels = labs_v)

  fin <- d8$value[is.finite(d8$value)]
  lo <- min(fin); hi <- max(fin); pad <- 0.08 * (hi - lo)

  p <- ggplot(mapping = aes(year, value)) +
    geom_point(data = si, colour = MUTED, size = 0.9, na.rm = TRUE) +
    geom_line(data = sf, colour = BLUE, linewidth = 0.6, na.rm = TRUE)
  if (nrow(rep)) {
    p <- p + geom_point(data = rep, colour = RED, size = 1.7, na.rm = TRUE)
  }
  p +
    facet_wrap(~panel, nrow = 1, scales = "free_x") +
    coord_cartesian(ylim = c(lo - pad, hi + pad)) +
    labs(title = title, subtitle = subtitle, x = NULL, y = NULL) +
    theme_x13() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),
          axis.line.x = element_blank(),
          panel.spacing.x = unit(4, "pt"))
}

# --------------------------------------------------------------------------- #
# 4. Forecast fan
# --------------------------------------------------------------------------- #
plot_forecast <- function(bundle, title, subtitle, history = 48) {
  base <- bundle_base(bundle)
  a1  <- read_save(bundle, "a1",  base)
  fct <- read_save(bundle, "fct", base)  # forecast, lowerci, upperci
  vn  <- attr(fct, "value_names")

  n <- nrow(a1)
  hist <- a1[max(1L, n - history + 1L):n, , drop = FALSE]
  cut <- hist$date[nrow(hist)]

  obs <- data.frame(date = hist$date, value = hist$value)
  fc  <- data.frame(date = fct$date,
                    forecast = fct[[vn[1]]],
                    lower = if (length(vn) >= 2) fct[[vn[2]]] else NA_real_,
                    upper = if (length(vn) >= 3) fct[[vn[3]]] else NA_real_)

  lvl <- c("Observed", "Forecast")
  p <- ggplot()
  if (all(is.finite(fc$lower)) && all(is.finite(fc$upper))) {
    p <- p + geom_ribbon(data = fc, aes(date, ymin = lower, ymax = upper,
                                        fill = "95% interval")) +
      scale_fill_manual(values = c("95% interval" = BLUE_LIGHT))
  }
  p +
    geom_vline(xintercept = as.numeric(cut), colour = BASELINE, linewidth = 0.4) +
    geom_line(data = obs, aes(date, value, colour = lvl[1]), linewidth = 0.76,
              na.rm = TRUE) +
    geom_line(data = fc, aes(date, forecast, colour = lvl[2]), linewidth = 0.76,
              linetype = "52", na.rm = TRUE) +
    scale_colour_manual(values = setNames(c(BLUE, BLUE_DARK), lvl),
                        breaks = lvl) +
    scale_x_date(expand = expansion(mult = c(0.01, 0.02))) +
    scale_y_continuous(labels = function(v) format(v, big.mark = ",",
                                                    scientific = FALSE,
                                                    trim = TRUE)) +
    labs(title = title, subtitle = subtitle, x = NULL, y = NULL) +
    guides(colour = guide_legend(order = 1,
                                 override.aes = list(linetype = c("solid", "52"))),
           fill = guide_legend(order = 2)) +
    theme_x13()
}

# --------------------------------------------------------------------------- #
# 5. M-statistics dashboard
# --------------------------------------------------------------------------- #
M_DESC <- c(
  m01 = "Irregular contribution",
  m02 = "Irregular in trend",
  m03 = "I/C ratio",
  m04 = "Irregular autocorrelation",
  m05 = "MCD months",
  m06 = "I/S yearly",
  m07 = "Moving seasonality",
  m08 = "Seasonal fluctuation",
  m09 = "Seasonal linear movement",
  m10 = "Recent seasonal fluct.",
  m11 = "Recent seasonal movement"
)

plot_mstats <- function(bundle, title, subtitle) {
  base <- bundle_base(bundle)
  udg  <- read_udg(bundle, base)
  keys <- sprintf("m%02d", 1:11)
  vals <- vapply(keys, function(k) udg_num(udg, paste0("f3.", k)), numeric(1))
  q <- udg_num(udg, "f3.q")

  df <- data.frame(
    key = keys,
    label = sprintf("%s  ·  %s", toupper(keys), M_DESC[keys]),
    value = vals,
    fail = !is.na(vals) & vals >= 1.0
  )
  # m01 at top -> reverse level order (ggplot puts first level at bottom).
  df$label <- factor(df$label, levels = rev(df$label))

  xmax <- max(3.2, max(vals, na.rm = TRUE) * 1.15)
  labpos <- df$value + 0.04 * xmax
  q_col <- if (!is.na(q) && q < 1.0) GOOD_TEXT else CRITICAL
  verdict <- if (!is.na(q) && q < 1.0) "accepted" else "not accepted"

  ggplot(df, aes(value, label)) +
    geom_col(aes(fill = fail), width = 0.62, na.rm = TRUE) +
    geom_vline(xintercept = 1.0, colour = INK2, linewidth = 0.5) +
    geom_text(aes(x = labpos, label = ifelse(is.na(value), "",
                                             sprintf("%.2f", value)),
                  colour = fail),
              hjust = 0, size = 3, na.rm = TRUE) +
    annotate("text", x = 1.02, y = length(keys) + 0.4, label = "acceptance limit",
             colour = INK2, size = 3, hjust = 0, vjust = 1) +
    annotate("text", x = xmax, y = length(keys) + 0.9,
             label = sprintf("Q = %.2f  (%s)", q, verdict),
             colour = q_col, fontface = "bold", size = 4, hjust = 1, vjust = 1) +
    scale_fill_manual(values = c(`FALSE` = BLUE, `TRUE` = CRITICAL),
                      guide = "none") +
    scale_colour_manual(values = c(`FALSE` = INK2, `TRUE` = CRITICAL),
                        guide = "none") +
    scale_x_continuous(limits = c(0, xmax), expand = expansion(mult = c(0, 0.02))) +
    coord_cartesian(clip = "off") +
    labs(title = title, subtitle = subtitle, x = NULL, y = NULL) +
    theme_x13(time_series = FALSE) +
    theme(axis.text.y = element_text(colour = INK, hjust = 0),
          axis.ticks.y = element_blank(),
          plot.margin = margin(12, 20, 10, 12))
}
